### ISSUE TITLE
Contao DB update fails to alter table tl_metamodel_levensthein_index

### ISSUE THREAD

**Issue by Sarahsolus** [MetaModels/attribute_levenshtein#5]
Since the last update, migrating or contao/install fails to do

ALTER TABLE tl_metamodel_levensthein_index CHANGE word word VARCHAR(64) CHARACTER SET utf8 DEFAULT '' NOT NULL COLLATE `utf8_bin`

the used database is utf8mb4_unicode_ci (MySQL 8.0)

metamodels v2.2
metamodels/attribute_levenshtein v2.2.2

**Comment by zonky2** [comment]
We cannot reproduce the error like this :(

What were the settings of the table or column before the migration - which MM version did you come from?

**Comment by Sarahsolus** [comment]
The DB is based on a Contao 4.4 - MM 2.1 version, that was upgraded to 4.9 - MM 2.2 running with MySQL5.7. That DB was imported into a MySQL8.0 (IONOS), with a fresh installation. After the first install the migration worked without problems. I did a composer update yesterday and afterwards the problem appeared. I also deleted the tl_metamodel_levensthein_index. It is recreated with the same problem.

**Comment by zonky2** [comment]
We also noticed this behaviour during an installation at Netcup.

The problem is that MM wants to create a column with `utf8_bin`. Contao has set `utf8mb4_bin` as the default and if the DB sees another variant, e.g. `utf8mb3_bin`, as the default, the DB "reports" everything OK but Contao expects something else. However, we cannot specify a `utf8mb4_bin` from MM... :(

You can check the settings of your DB as follows

```
SELECT @@character_set_database, @@collation_database;
```
or
```
SELECT DEFAULT_CHARACTER_SET_NAME, DEFAULT_COLLATION_NAME
FROM INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME = 'db_name';
```

You can force an `utf8mb4_bin` with your own DCA file

```
$GLOBALS['TL_DCA']['tl_metamodel_levensthein_index']['fields']['word']['sql'] =
    'varchar(64) COLLATE utf8mb4_bin NOT NULL default \'\'';
```

**Comment by Sarahsolus** [comment]
The DCA methode seems to work just fine. Thank you.

**Comment by amenk** [comment]
@zonky2 What is the intention of passing the collation in the DCA for tl_metamodel_levensthein_index ?

--- 
Very similar problem here, describe table shows, we are on a recent 2.3@dev with Contao 4.13 and MariaDB 10.6.12-MariaDB-0ubuntu0.22.04.1

```
  `word` varchar(64) CHARACTER SET utf8mb3 COLLATE utf8mb3_bin NOT NULL DEFAULT '',
```


Above solution sounds like a workaround? Is this eventually a bug? Not everybody should need to overwrite the DCA to get a clean DB update log?

Eventually related code in the core: https://github.com/contao/core-bundle/blob/e70a741b09c92b704ce4a754d92be4da2f9768a9/src/Doctrine/Schema/DcaSchemaProvider.php#L63

**Comment by discordier** [comment]
It is a workaround and, as a matter of fact, we found a permanent solution just the other day.

Please be patient until we have implemented it.

Until then, the work around helps.

**Comment by zonky2** [comment]
> Please be patient until we have implemented it.

MM 2.3

**Comment by amenk** [comment]
@zonky2 okay, but it's not yet in `2.3@dev`, I think.
Will have an eye on the MM2.3 newsletter :-)

*being-patient*

**Comment by zonky2** [comment]
Should fixed since ~3h - see commit number c3ff5b0d

**Comment by amenk** [comment]
@zonky2 I am happy to confirm the fix :-) Thank you very much
